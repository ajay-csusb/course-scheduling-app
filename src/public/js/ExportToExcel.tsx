import * as React from 'react';
import { IClass } from './Class';
import * as FileSaver from 'file-saver';
import * as Watchdog from './Watchdog';
import * as ClassSearchUtils from './ClassSearchUtils';

interface IExportToExcelProps {
  classes: IClass[];
}

export default class ExportToExcel extends React.Component<IExportToExcelProps>  {
  constructor(props: IExportToExcelProps) {
    super(props);
    this.exportToExcelOnClickHandler = this.exportToExcelOnClickHandler.bind(this);
  }

  render() {
    return (
      <div className="align-right">
        <button id="export-to-excel" className="btn-utility" onClick={this.exportToExcelOnClickHandler}>
          <i className="fas fa-file-download"></i>Export to Excel
        </button>
      </div>
    );
  }

  private exportToExcelOnClickHandler(): void {
    ClassSearchUtils.exportToExcelPost(this.props.classes)
    .then((blob: Blob) => {
      FileSaver.saveAs(blob, 'class-search.xlsx');
    })
    .catch((err: Error) => {
      Watchdog.log(err);
    });
  }

}
