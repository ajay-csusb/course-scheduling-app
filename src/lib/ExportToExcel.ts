
import { IClass, Class } from '../public/js/Class';
import * as ClassSearchUtils from '../public/js/ClassSearchUtils';
const excel = require('node-excel-export');

export interface ExcelData {
  campus: string;
  classNumber: number;
  classSection: string;
  courseAttr: string;
  csuUnits: number;
  fullSsrComponent: string;
  instructionMode: string;
  instructorName: string;
  meetingDay: string;
  meetingTime: string;
  seatsAvailable: string;
  sessionCode: string;
  subject: string;
  term: string;
  title: string;
  waitlistSeatsAvailable: string;
}

export function getExcelDocument(data: any) {
  return excel.buildExport([{
    name: 'Class Search Report',
    heading: getColumnTitle(),
    merges: getColumnMergeInfo(),
    specification: getColumnSpecification(),
    data: getDataForExcel(data),
  }]);
}

export function getDataForExcel(classes: IClass[]): ExcelData[] {
  const results: ExcelData[] = [];
  for (const _class of classes) {
    const result: ExcelData = extractExcelData(_class);
    results.push(result);
  }
  return results;
}

function extractExcelData(_class: IClass): ExcelData {
  const currClass: Class = new Class(_class);
  const { campus, catalogNo, classNumber, classSection, courseAttr, csuUnits, enrolledCapacity, fullSsrComponent, instructorName, quarter, subject, title, topic, waitlistCapacity} = _class;
  const result: ExcelData = {
    term: quarter,
    subject: `${subject} ${catalogNo}`,
    title: `${title}${ClassSearchUtils.formatSubjectTopic(topic)}`,
    campus: ClassSearchUtils.getCampusName(campus),
    classNumber: classNumber,
    classSection: classSection,
    courseAttr: courseAttr,
    csuUnits: csuUnits,
    fullSsrComponent: fullSsrComponent,
    instructionMode: ClassSearchUtils.getInstructionMode(_class),
    instructorName: instructorName,
    meetingDay: currClass.getClassMeetingDates(),
    meetingTime: currClass.getClassMeetingTimes(),
    seatsAvailable: `${ClassSearchUtils.getNoOfAvailableSeats(_class)}/${enrolledCapacity}`,
    sessionCode: ClassSearchUtils.getSessionCode(_class),
    waitlistSeatsAvailable: `${ClassSearchUtils.getNoOfAvailableSeatsInWaitlist(_class)}/${waitlistCapacity}`,
  };
  return result;
}

function getColumnSpecification() {
  const col: any = {};
  const colName = [
    'term',
    'subject',
    'title',
    'classSection',
    'classNumber',
    'fullSsrComponent',
    'csuUnits',
    'instructorName',
    'meetingDay',
    'meetingTime',
    'instructionMode',
    'sessionCode',
    'seatsAvailable',
    'waitlistSeatsAvailable',
    'courseAttr',
    'campus',
  ];
  for (const i of colName) {
    col[i] = {
      width: 120,
    };
  }
  return col;
}

function getColumnTitle() {
  return [
    [
      'Term',
      'Subject',
      'Title',
      'Section',
      'Class Number',
      'Types',
      'Unit',
      'Instructor',
      'Days(s)',
      'Time',
      'Mode',
      'Session',
      'Seats Available',
      'Wailtist Seats Available',
      'Attribute',
      'Campus',
    ],
  ];
}

function getColumnMergeInfo() {
  const results = [];

  for (let index = 1; index <= getColumnTitle()[0].length; index++) {
    results.push({ start: { row: 1, column: index }, end: { row: 2, column: index } });
  }

  return results;
}
