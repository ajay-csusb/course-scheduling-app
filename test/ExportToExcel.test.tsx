import ExportToExcel from '../src/public/js/ExportToExcel';
import { classJson, classPDC } from './ClassesJson';
import { mount, ReactWrapper, shallow } from 'enzyme';
import * as React from 'react';
import fetchMock from 'fetch-mock';
import { TestUtils } from './TestUtils';

describe('Export to Excel component', () => {
  let classes;

  beforeAll(() => {
    TestUtils.ajax();
    classes = [classJson, classPDC];
  });

  test('snapshot', () => {
    const exportToExcelComponent = <ExportToExcel classes={classes} />
    expect(exportToExcelComponent).toMatchSnapshot();
  });

  test('fetch parameters', (done) => {
    const exportToExcelComponent = mount(<ExportToExcel classes={classes} />);
    exportToExcelComponent.find('#export-to-excel').simulate('click');
    const fetchOptions = fetchMock.lastOptions();
    expect(fetchOptions.body).toMatch(JSON.stringify(classes));
    done();
  });

  test('markup', () => {
    const exportToExcelComponent = shallow(<ExportToExcel classes={classes} />);
    expect(exportToExcelComponent.html()).toContain('export-to-excel');
    expect(exportToExcelComponent.html()).toContain('fa-file-download');
    expect(exportToExcelComponent.html()).toContain('Export to Excel');
  });

});
