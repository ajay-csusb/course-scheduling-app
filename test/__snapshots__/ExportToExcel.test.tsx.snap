// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Export to Excel component snapshot 1`] = `
<ExportToExcel
  classes={
    Array [
      Object {
        "amount": 0,
        "buildingCode": "COE",
        "campus": "MAIN",
        "catalogNo": "101",
        "classEndTime": "1:50 PM",
        "classMeetingNo": 1,
        "classNumber": 101,
        "classSection": "01",
        "classStartTime": "12:00 PM",
        "classStatus": "Active",
        "courseAttr": "",
        "courseAttrDescription": "",
        "courseId": "000001",
        "csuUnits": 4,
        "degreeType": "PBAC",
        "description": "ACCT GOVT & NONPROFIT ORG",
        "employeeId": "001343367",
        "enrolledCapacity": 30,
        "enrolledTotal": 28,
        "enrollmentStatus": "",
        "facilityId": "JB0113",
        "fee": "0",
        "fri": "N",
        "fullSsrComponent": "",
        "geCourseAttr": "",
        "instructionMode": "P",
        "instructorName": "Dyck, Harold",
        "mon": "Y",
        "quarter": "2194",
        "room": "101",
        "sat": "N",
        "sessionCode": "1",
        "ssrComponent": "SEM",
        "subject": "ACCT",
        "sun": "N",
        "textbook": "",
        "thurs": "N",
        "title": "Introduction to Accounting",
        "topic": "",
        "tues": "N",
        "waitlistCapacity": 100,
        "waitlistTotal": 1,
        "wed": "Y",
      },
      Object {
        "amount": 0,
        "buildingCode": "PF",
        "campus": "PALM",
        "catalogNo": "201",
        "classEndTime": "7:50 PM",
        "classMeetingNo": 1,
        "classNumber": 102,
        "classSection": "01",
        "classStartTime": "6:00 PM",
        "classStatus": "Active",
        "courseAttr": "",
        "courseAttrDescription": "",
        "courseId": "000001",
        "csuUnits": 4,
        "degreeType": "EXED",
        "description": "ACCT GOVT & NONPROFIT ORG",
        "employeeId": "001343367",
        "enrolledCapacity": 30,
        "enrolledTotal": 28,
        "enrollmentStatus": "",
        "facilityId": "JB0113",
        "fee": "0",
        "fri": "N",
        "fullSsrComponent": "",
        "geCourseAttr": "",
        "instructionMode": "P",
        "instructorName": "Coates, Breena",
        "mon": "N",
        "quarter": "2192",
        "room": "101",
        "sat": "N",
        "sessionCode": "1",
        "ssrComponent": "SEM",
        "subject": "ECON",
        "sun": "N",
        "textbook": "",
        "thurs": "Y",
        "title": "",
        "topic": "",
        "tues": "Y",
        "waitlistCapacity": 100,
        "waitlistTotal": 1,
        "wed": "N",
      },
    ]
  }
>
  <div
    className="text-align-right"
  >
    <button
      className="btn-utility"
      id="export-to-excel"
      onClick={[Function]}
    >
      <i
        className="fas fa-file-download"
      />
      Export to Excel
    </button>
  </div>
</ExportToExcel>
`;
